src/org/nfrac/comportex/cells.cljc

```diff
@@ -666,42 +674,46 @@
                                           min-act :learn-threshold
                                           new-syns :new-synapse-count
                                           max-syns :max-synapse-count
-                                          max-segs :max-segments}]
-  (let [lci-vec (vec lci)] ;; for faster sampling
+                                          max-segs :max-segments
+                                          reuse-frac :relearning-similarity-frac}]
+  (let [any-reuse? (and reuse-frac (pos? reuse-frac) (> (count aci) (count lci)))
+        lci-seq (seq lci)] ;; for faster sampling
     (loop [cells (seq lc)
            m (transient {})
            rng rng]
       (if-let [cell-id (first cells)]
-        (if-let [seg-path (get well-matching-paths cell-id)]
-          ;; choose the well matching segment
-          (recur (next cells)
-                 (assoc! m cell-id (syn/seg-update seg-path :learn nil nil))
-                 rng)
-          ;; otherwise - not well matching - check disconnected synapses
-          (let [cell-segs (p/cell-segments sg cell-id)
-                [match-si exc seg] (best-matching-segment cell-segs aci
-                                                          min-act 0.0)
-                new-segment? (not match-si)
-                [seg-idx die-syns] (if match-si
-                                     [match-si nil]
-                                     (new-segment-id cell-segs pcon max-segs
-                                                     max-syns))
-                grow-n (- new-syns exc)
-                [rng* rng] (if (pos? grow-n) (random/split rng) [rng rng])
-                grow-source-ids (segment-new-synapse-source-ids seg lci-vec grow-n rng*)
-                die-source-ids (if new-segment?
-                                 (keys die-syns) ;; remove any existing (replaced)
-                                 (segment-excess-synapse-source-ids seg grow-n
-                                                                    max-syns))
-                seg-path (conj cell-id seg-idx)]
+        (let [[rng* rng] (if any-reuse? (random/split rng) [rng rng])
+              reuse? (when any-reuse? (< (random/rand-double rng*) reuse-frac))]
```

Comment:
You're right. Good point. I'll probably take your latter solution but for now I've removed the commit from this pull request. It doesn't belong here.